Add DeployAccount Method to account/account.go for Issue #381

Gabulhas opened the pull request:
Introduces a new DeployAccount method in account/account.go, aiming to simplify the process of account deployment on StarkNet as outlined in Issue #381. The DeployAccount method encapsulates multiple functions into a single call, making it easier and more intuitive for users to deploy accounts.

TODO:
- Add tests (I don't know if this is needed)
- Let users deploy accounts using the default UDC

Gabulhas committed:
created Account.DeployAccount method

Gabulhas committed:
added test for account.DeployAccount

Gabulhas committed:
Merge branch 'main' into feature/issue-382

rianhughes commented on account/account.go:
```diff
@@ -824,3 +824,49 @@ func FmtCalldataCairo2(fnCalls []rpc.FunctionCall) []*felt.Felt {
+	MaxFee              *felt.Felt
+	DeploytWaitTime     time.Duration
+	ConstructorCalldata []*felt.Felt
+}
```
Can you add comments (eg Classhash, ConstructorCallData). There's also a typo in DeploytWaitTime.

rianhughes commented on account/account.go:
```diff
@@ -824,3 +824,49 @@ func FmtCalldataCairo2(fnCalls []rpc.FunctionCall) []*felt.Felt {
+	ConstructorCalldata []*felt.Felt
+}
+
+func (account *Account) DeployAccount(options DeployOptions) (*rpc.AddDeployAccountTransactionResponse, error) {
```
Can we rename this to something like "CreateAndExecuteAddDeployAccount"?

Gabulhas replied:
Sure! 
The idea of just having "DeployAccount" was to be easier for the developer, since this is just a merge between two functions.

But will do that!

rianhughes commented on account/account.go:
```diff
@@ -824,3 +824,49 @@ func FmtCalldataCairo2(fnCalls []rpc.FunctionCall) []*felt.Felt {
+		return nil, err
+	}
+
+	resp, err := account.AddDeployAccountTransaction(context.Background(), tx)
```
You will need to wrap the tx in an rpc.BroadcastDeployAccountTxn type

Gabulhas replied:
Merged the upstream main branch with this PR.

rianhughes commented on account/account.go:
```diff
@@ -824,3 +824,49 @@ func FmtCalldataCairo2(fnCalls []rpc.FunctionCall) []*felt.Felt {
+		return nil, err
+	}
+
+	account.WaitForTransactionReceipt(context.Background(), resp.TransactionHash, options.DeploytWaitTime)
```
It's good practice to handle the return types, particularly the errors

rianhughes commented on account/account_test.go:
```diff
@@ -955,6 +955,50 @@ func TestAddDeclareTxn(t *testing.T) {
 	}
 }
 
+func TestDeployAccountDevnet(t *testing.T) {
```
This doesn't run for me (eg see wrapping tx type in rpc.BroadcastDeployAccountTxn comment above)

Gabulhas replied:
I need to merge the new RPCv0.5 with my branch, this applies to other parts of my code

rianhughes commented on examples/deployAccount/main.go:
```diff
@@ -1,9 +1,11 @@
 package main
```
Can you create a new example, I think it would be nice to have examples using both methods

rianhughes replied:
Also, this isnt running for me

Gabulhas replied:
Sure!

rianhughes commented on account/account_test.go:
```diff
@@ -955,6 +955,50 @@ func TestAddDeclareTxn(t *testing.T) {
+
+	deployOptions := account.DeployOptions{
+		ClassHash:           classHash,
+		MaxFee:              new(felt.Felt).SetUint64(0),
```
I think you might have to increase the MaxFee?

rianhughes commented on account/account_test.go:
```diff
@@ -955,6 +955,50 @@ func TestAddDeclareTxn(t *testing.T) {
+		ClassHash:           classHash,
+		MaxFee:              new(felt.Felt).SetUint64(0),
+		DeploytWaitTime:     2 * time.Second,
+		ConstructorCalldata: []*felt.Felt{},
```
Also, shouldn't this include "pub"?

rianhughes commented on examples/deployAccount/main.go:
```diff
@@ -59,39 +61,30 @@ func main() {
-	// Send transaction to the network
-	resp, err := acnt.AddDeployAccountTransaction(context.Background(), rpc.BroadcastDeployAccountTxn{DeployAccountTxn: tx})
+	fmt.Println("Waiting for deployment")
+	deployOptions := account.DeployOptions{
```
I think you need to set ConstructorCallData

rianhughes reviewed (changes requested):
I've left some comments.

Gabulhas committed:
Fixed typo and changed DeployAccount name

Gabulhas committed:
Merge remote-tracking branch 'origin/feature/issue-382' into feature/issue-382
Merge origin with feature

Gabulhas committed:
- Updated the TestDeployAccountDevnet func name to TestCreateAndExecuteAddDeployAccount
- Created a separate example for CreateAndExecuteAddDeployAccount
- Merged with upstream and updated CreateAndExecuteAddDeployAccount

(Fixed issues mentioned in NethermindEth/starknet.go/pull/445)

rianhughes commented on account/account.go:
```diff
@@ -824,3 +824,55 @@ func FmtCalldataCairo2(fnCalls []rpc.FunctionCall) []*felt.Felt {
+	ConstructorCalldata []*felt.Felt  // ConstructorCalldata contains the calldata to be passed to the constructor of the account contract upon deployment.
+}
+
+func (account *Account) CreateAndExecuteAddDeployAccount(options DeployOptions) (*rpc.AddDeployAccountTransactionResponse, error) {
```
This works for rpcv05 and lower. However, rpcv06 introduces a new deployAccount transaction version (v3), meaning users won't be able to use this method to submit v3 transactions. It would be great if we could make this compatible with both transaction versions (eg by accepting the transaction as an argument, tx rpc.DeployAccountType)

Gabulhas replied:
sure will fix it

Gabulhas replied:
by "this method" you mean the whole method "CreateAndExecuteAddDeployAccount" or the way you deploy accounts like in the function?

rianhughes replied:
The CreateAndExecuteAddDeployAccount method.
